plugins/plugin-bitbucket/codenvy-plugin-bitbucket-server-webhooks/src/test/java/com/codenvy/plugin/webhooks/bitbucketserver/BitbucketServerWebhookServiceTest.java

```diff
@@ -0,0 +1,141 @@
+import com.codenvy.plugin.webhooks.bitbucketserver.shared.User;
+
+import org.eclipse.che.api.core.rest.shared.dto.Link;
+import org.eclipse.che.api.factory.shared.dto.FactoryDto;
+import org.eclipse.che.api.workspace.shared.dto.ProjectConfigDto;
+import org.eclipse.che.api.workspace.shared.dto.SourceStorageDto;
+import org.eclipse.che.api.workspace.shared.dto.WorkspaceConfigDto;
+import org.mockito.testng.MockitoTestNGListener;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+import javax.servlet.ServletInputStream;
+import javax.servlet.http.HttpServletRequest;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.util.Collections.singletonList;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+
+@Listeners(value = MockitoTestNGListener.class)
+public class BitbucketServerWebhookServiceTest {
+
+    private BitbucketServerWebhookService service;
+
+    private final Map<String, String> parameters = new HashMap<>();
+
+    @BeforeMethod
+    public void setup() throws Exception {
+        FactoryConnection factoryConnection = mock(FactoryConnection.class);
+        FactoryDto factory = mock(FactoryDto.class);
+        WorkspaceConfigDto workspace = mock(WorkspaceConfigDto.class);
+        ProjectConfigDto project = mock(ProjectConfigDto.class);
+        SourceStorageDto source = mock(SourceStorageDto.class);
+
+        when(factory.getWorkspace()).thenReturn(workspace);
+        when(factory.getLink(anyString())).thenReturn(mock(Link.class));
+        when(factory.getId()).thenReturn("factoryId");
+        when(factoryConnection.getFactory("factoryId")).thenReturn(factory);
+        when(factoryConnection.updateFactory(factory)).thenReturn(factory);
+        when(workspace.getProjects()).thenReturn(singletonList(project));
+        when(project.getSource()).thenReturn(source);
+        when(source.getType()).thenReturn("type");
+        when(source.getLocation()).thenReturn("http://owner@bitbucketserver.host/scm/projectkey/repository.git");
+        parameters.put("branch", "testBranch");
+        when(source.getParameters()).thenReturn(parameters);
+
+        service = spy(new BitbucketServerWebhookService(mock(AuthConnection.class), factoryConnection, "http://bitbucketserver.host/"));
+    }
+
+    @Test
+    public void shouldHandlePushEvent() throws Exception {
+        //given
+        PushEvent event = newDto(PushEvent.class)
```

Comment:
I'd avoid using of static import in this case and format this block code of code like
```
DtoFactory.newDto(PushEvent.class)
          .withRepository(DtoFactory.newDto(Repository.class)
                                    .withProject(DtoFactory.newDto(Project.class)
                                                           .withOwner(DtoFactory.newDto(User.class)
                                                                                .withName("owner"))
                                                           .withKey("projectkey"))
                                    .withName("repository"))
          .withRefChanges(singletonList(DtoFactory.newDto(RefChange.class)
                                                  .withToHash("hash commit")
                                                  .withType("UPDATE")
                                                  .withRefId("refs/heads/testBranch")))
          .withChangesets(DtoFactory.newDto(Changesets.class)
                                    .withValues(singletonList(DtoFactory.newDto(Changeset.class)
                                                                        .withToCommit(DtoFactory.newDto(Commit.class)
                                                                                                .withId("hash commit")
                                                                                                .withMessage("message")))));
```
IMHO It makes code clear
Also you use this very difficult for understanding code twice. It'd be better to move it to private method `createPushEvent(String message)`